plugins/meta/vrf/vrf.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+)
+
+// findVRF finds a VRF link with the provided name.
+func findVRF(name string) (*netlink.Vrf, error) {
+	link, err := netlink.LinkByName(name)
+	if err != nil {
+		return nil, err
+	}
+	vrf, ok := link.(*netlink.Vrf)
+	if !ok {
+		return nil, fmt.Errorf("Netlink %s is not a VRF", name)
+	}
+	return vrf, nil
+}
+
+// createVRF creates a new VRF and sets it up.
+func createVRF(name string) (*netlink.Vrf, error) {
+	tableID, err := findFreeRoutingTableID()
+	if err != nil {
+		return nil, err
+	}
+	vrf := &netlink.Vrf{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: name,
+		},
+		Table: tableID,
+	}
+
+	err = netlink.LinkAdd(vrf)
+	if err != nil {
+		return nil, fmt.Errorf("could not add VRF %s: %v", name, err)
+	}
+	err = netlink.LinkSetUp(vrf)
+	if err != nil {
+		return nil, fmt.Errorf("could not set link up for VRF %s: %v", name, err)
+	}
+
+	return vrf, nil
+}
+
+// assignedInterfaces returns the list of interfaces associated to the given vrf.
+func assignedInterfaces(vrf *netlink.Vrf) ([]netlink.Link, error) {
+	links, err := netlink.LinkList()
+	if err != nil {
+		return nil, fmt.Errorf("getAssignedInterfaces: Failed to find links %v", err)
+	}
+	res := make([]netlink.Link, 0)
+	for _, l := range links {
+		if l.Attrs().MasterIndex == vrf.Index {
+			res = append(res, l)
+		}
+	}
+	return res, nil
+}
+
+// addInterface adds the given interface to the VRF
+func addInterface(vrf *netlink.Vrf, intf string) error {
+	i, err := netlink.LinkByName(intf)
+	if err != nil {
+		return fmt.Errorf("could not get link by name %s", intf)
+	}
+
+	// IPV6 addresses are not maintained unless
+	// sysctl -w net.ipv6.conf.all.keep_addr_on_down=1 is called
+	// so we save it, and restore it back.
+	addresses, err := netlink.AddrList(i, netlink.FAMILY_V6)
+	if err != nil {
+		return fmt.Errorf("failed getting ipv6 addresses for %s", intf)
+	}
+	err = netlink.LinkSetMaster(i, vrf)
+	if err != nil {
+		return fmt.Errorf("could not set vrf %s as master of %s: %v", vrf.Name, intf, err)
+	}
+
+	for _, addr := range addresses {
+		err = netlink.AddrAdd(i, &addr)
+		if err != nil {
+			return fmt.Errorf("could not restore address %s to %s @ %s: %v", addr, intf, vrf.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func findFreeRoutingTableID() (uint32, error) {
+	var maxTable uint32
+	links, err := netlink.LinkList()
+	if err != nil {
+		return 0, fmt.Errorf("findFreeRoutingTableID: Failed to find links %v", err)
+	}
+	for _, l := range links {
+		if vrf, ok := l.(*netlink.Vrf); ok {
+			if vrf.Table > maxTable {
+				maxTable = vrf.Table
+			}
+		}
+	}
+	return (maxTable + 1), nil
```

Comment:
routing table id seems to be unsigned char, uint8 in golang. Could you please double check it?
https://github.com/torvalds/linux/blob/master/include/uapi/linux/rtnetlink.h#L227

Reply:
Sure, thanks for pointing out!

Reply:
@s1061123  I _think_ that's a u32.

See https://elixir.bootlin.com/linux/latest/source/drivers/net/vrf.c#L44
https://elixir.bootlin.com/linux/latest/source/drivers/net/vrf.c#L1339

Which also sticks with the type used by the netlink library: https://github.com/containernetworking/plugins/blob/53377543e1e91e984033e0af5e19bc9bb6973c65/vendor/github.com/vishvananda/netlink/link_linux.go#L2745